### PR TITLE
Multi-selection for upload field in gallery

### PR DESCRIPTION
**Pull request by flamerohr**
Please see https://github.com/silverstripe/silverstripe-asset-admin/issues/693 for more details

Requires https://github.com/silverstripe/silverstripe-admin/pull/450 to be merged first

### CODE REVIEW

File: client/src/components/UploadField/UploadField.js [hunk 1]
```diff
@@ -43,31 +64,22 @@ class UploadField extends Component {
...
-    // Check length
-    if (left.length !== right.length) {
-      return true;
-    }
-    // Check ids appear in the same order
-    for (let i = 0; i < left.length; i++) {
-      if (left[i].id !== right[i].id) {
-        return true;
-      }
+  getMaxFiles() {
```
Comment (tractorcow) [inline client/src/components/UploadField/UploadField.js]: Do we need to bind these new methods in the constructor also?

File: client/src/components/UploadField/UploadField.js [hunk 2]
```diff
@@ -196,11 +208,22 @@ class UploadField extends Component {
    */
   handleAddInsert(data, file) {
     this.props.actions.uploadField.addFile(this.props.id, file);
-    this.handleAddHide();
+    this.handleHide();
 
     return Promise.resolve({});
   }
 
+  handleInsertMany(files) {
```
Comment (tractorcow) [inline client/src/components/UploadField/UploadField.js]: jsdoc ;)

**Comment by tractorcow** [comment]
All the travis behat errors match my own local test.

```
    tests/behat/features/manage-file-permissions.feature:16
    tests/behat/features/manage-file-permissions.feature:36
    tests/behat/features/manage-files.feature:49
    tests/behat/features/manage-files.feature:82
    tests/behat/features/manage-files.feature:101
```

:D Will fix now.

**Comment by tractorcow** [comment]
Edit batch action

image

**Comment by tractorcow** [comment]
Ok fixed but needs @flamerohr review on my tweaks to refactoring in Gallery.js

**Comment by tractorcow** [comment]
As a note, we've decided to do a bit of refactoring, ad shift `combinedFiles` into `files` property, and instead getting rid of the separate categories.

@flamerohr will push up the refactor later.

**Comment by flamerohr** [comment]
Yup, should have something up by tomorrow, got a bit done but wanted to test before pushing up

File: client/src/components/GalleryItem/GalleryItem.js [hunk 3]
```diff
@@ -255,7 +256,7 @@ class GalleryItem extends Component {
    * @returns {boolean}
    */
   uploading() {
-    return this.props.item.uploading;
```
Comment (tractorcow) [inline client/src/components/GalleryItem/GalleryItem.js]: We need a flag for items that can be activated, I think. how about `ready() { return this.props.item.id > 0; }`

File: client/src/components/GalleryItem/GalleryItem.js [hunk 4]
```diff
@@ -22,7 +22,8 @@ function shouldLoadImage(props) {
   return props.item.thumbnail
     && props.item.category === 'image'
     && props.item.exists
-    && !props.item.uploading // Don't load images for uploaded images (retain client thumbnail)
+    // Don't load images for uploaded images (retain client thumbnail)
+    && !(props.item.queuedId && props.item.progress < 100)
```
[the remark this answers is not in the document]
Comment (flamerohr) [reply client/src/components/GalleryItem/GalleryItem.js]: oops yeah, `queuedId` is enough :)

File: client/src/components/UploadField/tests/UploadFieldItem-test.js [hunk 5]
```diff
@@ -132,6 +133,7 @@ describe('UploadFieldItem', () => {
 
 
     it('does not display progress bar for errors', () => {
+      props.item.progress = 100;
```
Comment (flamerohr) [inline client/src/components/UploadField/tests/UploadFieldItem-test.js]: there are two progresses here ;)

Comment (tractorcow) [reply client/src/components/UploadField/tests/UploadFieldItem-test.js]: That's how progressive I am.

**Comment by tractorcow** [comment]
I'll check out the new look and feel tomorrow. Looks good so far though.

**Pull request merged**